### PR TITLE
Refactor, Replace Bootstrap with Bulma

### PR DESCRIPTION
**Pull request by juancarlospaco**
Refactor, see https://github.com/ThomasTJdev/nim_websitecreator/issues/22#issuecomment-429519107 and https://github.com/ThomasTJdev/nim_websitecreator/issues/23#issuecomment-429326459
Clean out.

**Frontend Help welcome.**
**Wording Grammar Review welcome.**

WIP

temp

temp

temp

temp

temp

:cat:

### CODE REVIEW

**Comment by ThomasTJdev** [comment]
Hi @juancarlospaco 
I have been testing the new layout and it looks good. I have a rough list below, which I will start to look at in prioritized order (OK?):
- I have some trouble with the arrangement of elements - (row/col/alignement)
- Some CSS glitches
- Colors of the login page
- Missing standard text sizes for fonts
- A base CSS formatting for blogpost
- The main hurdle is, that jQuery is required on the admin pages - especially when editing pages and blog. Either rewrite js_admin.js or append jQuery for all admin pages.
- Rewriting JS to avoid jQuery on non-admin pages

## Pictures
Here's how it looks:

### Settings
nimwc_arrangement
____
### Blog edit layout
nimwc_blogeditlayout
____
### Blog layout
nimwc_bloglayout
____
### Frontpage
nimwc_frontpage
____
### Edit main area
nimwc_editmainarea
____
### Navbar (logged in)
nimwc_navbar

**Comment by juancarlospaco** [comment]
Append jQuery for all admin pages and rewrite JS to avoid jQuery on non-admin pages.

Make sure you try on Incognito mode, sometimes browser cache gets your UI weird too.

Fontpage you said you are gonna replace it with more Documentation oriented content, so I didnt touch it.

**Comment by juancarlospaco** [comment]
Ive manually tested all of it, it looks and feels really cool :grey_exclamation:  :tada::smiley_cat::+1:

**Comment by juancarlospaco** [comment]
Can `urlEncoderCustom()` be replaced with `uri.encodeUrl( usePlus = true)` :grey_question: 
https://github.com/ThomasTJdev/nim_websitecreator/blob/master/nimwcpkg/resources/web/urltools.nim#L5
Or I am missing something... :thinking:

**Comment by ThomasTJdev** [comment]
> Can `urlEncoderCustom()` be replaced with `uri.encodeUrl( usePlus = true)` 
> 
> * https://github.com/ThomasTJdev/nim_websitecreator/blob/master/nimwcpkg/resources/web/urltools.nim#L5
> * https://nim-lang.org/docs/uri.html#encodeUrl%2Cstring
> 
> Or I am missing something...

I think my original idea for the custom implementation was:
1) To ensure pretty URLs
2) Escape special characters in URL ()*#?`'^~œæøåôõâö etc.

```nim
echo urlEncoderCustom("url with hashtag#") # ==> url+with+hashtag
echo encodeUrl("url with hashtag#", true) # ==> url+with+hashtag%23
```

Should we allow the use of special chars which anyways will be converted to e.g. %23? I like to give the freedom to the user, but I dont like ugly URLs ;-)

**Comment by juancarlospaco** [comment]
Some spoken languages need those I think, like Arab, Portuguese, Japanese, etc.

**Comment by ThomasTJdev** [comment]
> Some spoken languages need those I think, like Arab, Portuguese, Japanese, etc.

My own included ;) 

IDN's are supported using [Punycode](https://en.wikipedia.org/wiki/Punycode) but how should we handle it? Save it using `encodeUri`, convert to IDN directly (æble.dk ==> xn--ble-xla.dk) or save it with clear text and let the browser handle the parsing?
I have had some trouble with how to show the URL's when generating them as link and for the user to edit them again.

But I'm up for a another approach.

**Comment by juancarlospaco** [comment]
I think we should not make our own standard just for this, I say keep it simple, I see 2 options:

Use the std lib `encodeUrl()` and use the `pattern` to only allow ASCII of `<input>` tag on the URL.
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/url#Pattern_validation

Use Punycode, we also have Punycode on std lib because Nim is awesome.

**Comment by juancarlospaco** [comment]
- Super minimal Log Viewer.
- Implemented auto-rotating file logging, the option to notify Admin still exists on a `template`.

**Comment by juancarlospaco** [comment]
Honestly the mention on readme dont adds too much for me, but Collaborator can be cool, for example lets say you are on holidays travelling and NimWC explodes I can at least fix it right away. 
:smile:

**Comment by juancarlospaco** [comment]
https://mathiasbynens.github.io/rel-noopener

**Comment by juancarlospaco** [comment]
We dont have Login / Logout button on Mobile.
Can you please take a look at that before we Merge ?.

**Comment by juancarlospaco** [comment]
temporalote

**Comment by juancarlospaco** [comment]
i-ship-it-gif-9

**Comment by ThomasTJdev** [comment]
Merge-time this weekend :) !?

- urlEncoderCustom()
- Login button on mobile

Other stuff before merging?

**Comment by juancarlospaco** [comment]
I run out of improvements for current iteration of changes, everything looks good, 84 Commits so far. :+1:

File: nimwcpkg/nimwc_main.nim [hunk 1]
```diff
@@ -165,23 +155,22 @@ macro extensionCss(): string =
...
   let mainDir = replace(dir, "nimwcpkg", "")
 
-
   var extensions = ""
   for ppath in pluginsPath:
     let splitted = split(ppath, "/")
 
     if staticRead(ppath & "/public/style.css") != "":
-      discard staticExec("cp " & ppath & "/public/style.css " & mainDir & "/public/css/" & splitted[splitted.len-1] & ".css")
+      moveFile(source=ppath & "/public/style.css",
```
Comment (ThomasTJdev) [inline nimwcpkg/nimwc_main.nim]: moveFile is not supported in macro:
```nim
../../../../.choosenim/toolchains/nim-#head/lib/pure/os.nim(669, 8) Error: cannot 'importc' variable at compile time
```

I'm rolling back to staticExec for now.

Comment (ThomasTJdev) [inline nimwcpkg/nimwc_main.nim]: moveFile() is not supported in macro:
```nim
tmpl/main.tmpl(109, 24) template/generic instantiation of `extensionCss` from here
../../../../.choosenim/toolchains/nim-#head/lib/pure/os.nim(669, 8) Error: cannot 'importc' variable at compile time
```

I'm rolling back to staticExec for now.

**Comment by juancarlospaco** [comment]
Cool :grey_exclamation: 
Make a new Release so Nimble catch up the version change: https://github.com/ThomasTJdev/nim_websitecreator/releases

**Comment by ThomasTJdev** [comment]
Thank you very much for the tremendous work! NimWC is really nice now :)

I'm making you a collaborator now. Code review is required for breaking changes and design changes.

**Comment by juancarlospaco** [comment]
The Demo at https://nimwc.org is using the old version, needs Re-Deploy.

File: nimwcpkg/tmpl/sitemap.tmpl [hunk 2]
```diff
@@ -7,7 +7,27 @@
 #
 #proc genSitemap(): string =
 # result = ""
+# when defined(release):
```
Comment (ThomasTJdev) [inline nimwcpkg/tmpl/sitemap.tmpl]: Is there a reason for using defined(release). I can't seem to find a difference between the code in the if-statement.

Comment (juancarlospaco) [reply nimwcpkg/tmpl/sitemap.tmpl]: Its Minified for performance, but feel free to remove it if you dont like it, its not very important at all. :sweat_smile:

Comment (ThomasTJdev) [reply nimwcpkg/tmpl/sitemap.tmpl]: Ahh now I see it :smiley: ! I think I'll remove it to keep the code simple.

**Comment by ThomasTJdev** [comment]
> I think we should not make our own standard just for this, I say keep it simple, I see 2 options:
> 
> Use the std lib `encodeUrl()` and use the `pattern` to only allow ASCII of `<input>` tag on the URL.
> https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/url#Pattern_validation
> 
> Use Punycode, we also have Punycode on std lib because Nim is awesome.

I ran into trouble..

I am naming my blog posts:
website.com/blog/YYYY/MM/DD/name

But this now returns the following in the addressbar in Chrome.
https://website.com/blog/2018%2F11%2F16%2Fmy-normal-name

**Pull request merged**